### PR TITLE
Remove from disk: add to deck track menu, fix undeleted message

### PR DESCRIPTION
**Pull request by ronso0**
I was missing Remove From Disk in the deck menu when I was too lazy to look for the track in the library.
Here it is.

### CODE REVIEW

File: src/widget/wtrackmenu.cpp
Comment (ronso0) [inline src/widget/wtrackmenu.cpp:543]: I hope I got that right...

File: src/widget/wtrackmenu.cpp
Comment (ronso0) [inline src/widget/wtrackmenu.cpp:1810]: I'm not sure this is the best way to do it, but the first and simplest that came to mind.

Comment (ronso0) [reply src/widget/wtrackmenu.cpp:1810]: Do we need to check if it's the deleted track that's loaded?
The progress dialog is modal so manual unloading is not possible in he GUI, but with keyboards and controllers.

File: src/widget/wtrackmenu.cpp
Comment (uklotzde) [reply src/widget/wtrackmenu.cpp:543]: You don't want an early return in this case. Also `m_pTrackModel != nullptr` is not checked for any of the other options. It should be checked only once earlier if needed.

What was wrong with the existing code?

File: src/widget/wtrackmenu.cpp
Comment (uklotzde) [inline src/widget/wtrackmenu.cpp:797]: Why is this check needed here but not for any other option?

File: src/widget/wtrackmenu.cpp
Comment (uklotzde) [inline src/widget/wtrackmenu.cpp:1838]: Is this code reachable? How if `m_pTrackModel` equals `nullptr`?

**Review by uklotzde — CHANGES_REQUESTED** [review]

File: src/widget/wtrackmenu.cpp
Comment (uklotzde) [reply src/widget/wtrackmenu.cpp:1810]: Deleting an MP3 file while the audio stream is decoded might crash Mixxx: https://bugs.launchpad.net/mixxx/+bug/1452005

This happens because libmad/SoundSourceMP3 uses memory mapped files. You are not able to detect how many decoders are active. Checking if a track is loaded on a deck is not sufficient.

Better drop libmad/SoundSourceMP3 and use something more robust.

File: src/widget/wtrackmenu.cpp
Comment (ronso0) [reply src/widget/wtrackmenu.cpp:543]: My assumption when looking at how the track menu is initialized for either `WTrackproperty`, `WTrackWidgetGroup` or `WTrackTableView` is that we can end up here only if either there's a track model or a track loaded. So no track model = track loaded, right?
So the existing code would add the action to the track widgets' menu.

:facepalm: oouh I'm so shortsighted at this time of the day.. of course I don't want to quit `setupActions()`
will fix

File: src/widget/wtrackmenu.cpp
Comment (ronso0) [reply src/widget/wtrackmenu.cpp:797]: Based on the assumption outlined above:
either track loaded or track model (and track refs loaded), so if the menu is initialized from a track widget then there's no track model. > segfault

File: src/widget/wtrackmenu.cpp
Comment (ronso0) [reply src/widget/wtrackmenu.cpp:1810]: This might also happen with the existing delete & purge action, right? Like when the Analysis feature is busy.

How do you propose to proceed?

File: src/widget/wtrackmenu.cpp
Comment (ronso0) [reply src/widget/wtrackmenu.cpp:543]: > What was wrong with the existing code?

Ha :bulb: I didn't question my own code you already reviewed ; ]
`featureIsEnabled()` already checks `m_pTrackModel->hasCapabilities()` (if the feature requires a track model) so that was redundant in the first place...

**Comment by ronso0** [comment]
I added another check and removed the redundant feature test.
I'll squash the fixup as soon as I get thumbs up.

File: src/widget/wtrackmenu.cpp
Comment (uklotzde) [reply src/widget/wtrackmenu.cpp:1835]: Confused the different modes. Ignore my comment.

**Comment by uklotzde** [comment]
On second thought we should **not** enable the option to delete the file of a track that is loaded on a deck. This option should never be enabled if m_pTrack is set.

**Comment by Swiftb0y** [comment]
> It is counterintuitive to delete the file of a track while it is still loaded. Even if it doesn't crash Mixxx.

I disagree to be honest. Its counterintuitive if I want to delete a track but can't for no apparent reason. Maybe I just listened to the track and decided that it doesn't belong into my library, so I want to delete it but can't because its still loaded in a deck.

**Comment by Swiftb0y** [comment]
If we can't delete a track while its loaded for technical reason, I'd make that transparent so the user is confused (display the option, but grey it out and explain in a tooltip or something).

**Comment by ronso0** [comment]
we can simply unload it before deleting the file.

**Comment by ronso0** [comment]
Marked as draft because --in case it's decided to allow the operation for decks-- I'd squash the fixup commits.

This is the updated warning (for decks):
Screenshot_2021-12-09_23-22-09

File: src/widget/wtrackmenu.cpp
Comment (uklotzde) [inline src/widget/wtrackmenu.cpp:1781]: This code drags dependencies into WTrackMenu that don't belong here. Another example of the spaghetti control flow in Mixxx, unmaintainable.

Is this side-effect even synchronous? Can we guarantee that the track is ejected immediately? Probably not, because decoding is done concurrently in a separate thread -> CachingReaderWorker.

I suppose this doesn't work as intended.

Comment (ronso0) [reply src/widget/wtrackmenu.cpp:1781]: This is **only** supposed to have a clean UX: it would be confusing if the deleted track was still loaded and could still be played.

Please clarify your technical concerns (for example the libmad issue) and give an estimation if this can be solved at all (by me, with guidance, of course). Because, if there are issues to be expected they apply to the delete and purge action in general, not just to the deck option, since we're not checking if a track queued for deletion is analyzed or loaded anywhere. If there are serious concerns it'd be sad they pop up just now because we'd then have wasted quite some time implementing the feature in the first place :\ (what I've learned along the way would be persist, but still..)

Comment (uklotzde) [reply src/widget/wtrackmenu.cpp:1781]: I can't imagine an easy fix considering the current state of the code base. Merging the code for deleting files directly from within Mixxx without any chance to detect if those files are still accessed by Mixxx probably was a mistake. We didn't consider all the consequences.

I don't veto against merging this PR, but on the other hand I won't merge it myself. Someone else may decide.

**Review by uklotzde — APPROVED** [review]

File: src/widget/wtrackmenu.cpp
Comment (ronso0) [reply src/widget/wtrackmenu.cpp:1781]: Hmm, that's unfortunate.
I'm puzzled why you approve the changes even though you have serious technical objections.

I still don't overlook the potential consequences:
The worst case would be Mixxx crashing when deleting mp3 files due to the libmad / SoundSourceMP3 / mmap issue, right? In that case I wouldn't mind removing the entire feature.
Besides that, what could happen that harms the database or puts Mixxx into an inconsistent state?

Comment (uklotzde) [reply src/widget/wtrackmenu.cpp:1781]: I could only delete the entire review or keep the rejection. Both are not an option for me. "Disagree and merge" or "Disagree and let others merge" (for a second opinion).

I don't really care enough and I also don't have a better proposal. The code is just ugly with the entanglement of the control proxies for triggering a side-effect in another component directly. It works somehow, but probably not always.

Comment (ronso0) [reply src/widget/wtrackmenu.cpp:1781]: Okay, then thanks for your review!
I'll leave to others to judge the remove feature.

**Comment by ronso0** [comment]
(I squashed the small fixup commits)

File: src/widget/wtrackmenu.cpp
Comment (daschuer) [reply src/widget/wtrackmenu.cpp:792]: Please add a source code comment about it.

File: src/widget/wtrackmenu.cpp
Comment (daschuer) [inline src/widget/wtrackmenu.cpp:1735]: This comment is pointless now.

File: src/widget/wtrackmenu.cpp
Comment (daschuer) [inline src/widget/wtrackmenu.cpp:1739]: This and below can become a QStringLiteral

File: src/widget/wtrackmenu.cpp
Comment (daschuer) [inline src/widget/wtrackmenu.cpp:1742]: On Linux you can delete the file while playing.
It is actually removed after the last reference to the file has been closed.
On Windows the file is locked until all file handles are closed. 
I am not sure how it relates to caching. The file is kept in cache after rejecting. 
We need to check if the lock is released than.

File: src/widget/wtrackmenu.cpp
Comment (daschuer) [inline src/widget/wtrackmenu.cpp:1785]: You can use the static ControlObject function here:
ControlObject::set(ConfigKey(m_deckGroup, "stop"), 1.0);

File: src/widget/wtrackmenu.cpp
Comment (daschuer) [inline src/widget/wtrackmenu.cpp:1851]: Do we also shoe the not deleted track file? 
If yes, the text needs to reflect that.

**Review by daschuer — CHANGES_REQUESTED** [review]
Thank you. I have some final comment:

File: src/widget/wtrackmenu.cpp
Comment (ronso0) [reply src/widget/wtrackmenu.cpp:1735]: This and the comments below are meant to structure the dialog setup.
I removed this on because the string name is directly below.

File: src/widget/wtrackmenu.cpp
Comment (ronso0) [reply src/widget/wtrackmenu.cpp:1785]: :+1: Thanks

File: src/widget/wtrackmenu.cpp
Comment (ronso0) [reply src/widget/wtrackmenu.cpp:1851]: Yes, remaining file(s) are listed in the dialog.
I reworded the single file message to match the multiple files message.

File: src/widget/wtrackmenu.cpp
Comment (ronso0) [reply src/widget/wtrackmenu.cpp]: done.

**Review by daschuer — APPROVED** [review]

**Comment by daschuer** [comment]
LGTM, Thank you. 

Sorry for the delay, this one has slipped through.

**Comment by ronso0** [comment]
Never mind, this one wasn't urgent.
Did you have chance to test it on windows?

**Comment by daschuer** [comment]
Yes, at least on my Win 10 Virtual Machine it works without any issues.

**Pull request merged**
